samples/msal-node-samples/EndToEndTestApp/app/test/testUtils.ts

```diff
@@ -0,0 +1,119 @@
+    await page.waitForSelector("#idSIButton9");
+    await page.click("#idSIButton9");
+    await takeScreenshotAfter(6000, screenshot, page, 'consentApproved'); 
+}
+
+export async function clickSignIn(page: Page, screenshot: Screenshot): Promise<void> {
+    await screenshot.takeScreenshot(page, "samplePageInit");
+    await page.click("#SignIn");
+    await screenshot.takeScreenshot(page, "signInClicked");
+}
+
+export async function enterCredentialsADFS(page: Page, screenshot: Screenshot, username: string, accountPwd: string): Promise<void> {
+    await page.waitForSelector("#i0116");
+    await screenshot.takeScreenshot(page, "loginPageADFS");
+    await page.type("#i0116", username);
+    await page.click("#idSIButton9");
+    await page.waitForSelector("#userNameInput");
+    await screenshot.takeScreenshot(page, "adfsUsernameInputPage");
+    await page.type("#passwordInput", accountPwd);
+    await page.click("#submitButton");
+    await takeScreenshotAfter(6000, screenshot, page, `pwdSubmitted`);
+}
+
+export async function enterDeviceCode(page: Page, screenshot: Screenshot, code: string, deviceCodeUrl: string): Promise<void> {
+    await page.goto(deviceCodeUrl);
+    await page.waitForSelector("#otc");
+    await screenshot.takeScreenshot(page, 'deviceCodePage');
+    await page.type("#otc", code);
+    await page.click("#idSIButton9");
+}
+
+export function extractDeviceCodeParameters(output: string): { deviceCode: string, deviceLoginUrl: string } | null {
+    const deviceLoginUrlRegex = /(?<url>(https?|chrome):\/\/[^\s$.?#].[^\s]*)/;
+    const deviceCodeRegex = /code (?<code>\w+)/
+
+    if (deviceCodeRegex.test(output) && deviceLoginUrlRegex.test(output)) {
+        const deviceCode = output.match(deviceCodeRegex).groups.code;
+        const deviceLoginUrl = output.match(deviceLoginUrlRegex).groups.url;
+
+        return {
+            deviceCode,
+            deviceLoginUrl,
+        };
+    }
+
+    return null;
+}
+
+export function takeScreenshotAfter(duration: number, screenshot: Screenshot, page: Page, label: string): Promise<void> {
+    return new Promise(resolve => setTimeout(() => screenshot.takeScreenshot(page, label).then(() => resolve()), duration));
+}
+
+export async function validateCacheLocation(cacheLocation: string): Promise<void> {
+    return new Promise((resolve, reject) => {
+        fs.readFile(cacheLocation, "utf-8", (err, data) => {
+            if (err || data === "") {
+                fs.writeFile(cacheLocation, "{}", (error) => {
+                    if (error) {
+                        console.log("Error writing to cache file: ", error);
+                        reject();
+                    } else {
+                        resolve();
+                    }
+                });
+            } else {
+                resolve();
+            }
+        });
+    });
+}
+
+export async function sleep(delay: number) {
```

Comment:
Would prefer not to have arbitrary sleeps in our tests.

Reply:
Removed this function and it's use cases in the test suite.